### PR TITLE
Voice of god refactor; no more 'else if' cascade.

### LINKED ISSUE #33650: Runtime when AI uses Voice of God spell
> [11:52:56] Runtime in ,: bad arg name 'spans'
>   proc name: say (/mob/living/silicon/ai/say)
>   usr: Emerald (coiax) (/mob/living/silicon/ai)
>   usr.loc: The floor (171,134,2) (/turf/open/floor/plasteel/neutral/corner)
>   src: Emerald (/mob/living/silicon/ai)
>   src.loc: the floor (171,134,2) (/turf/open/floor/plasteel/neutral/corner)
>   call stack:
>   Emerald (/mob/living/silicon/ai): say("HI THERE", null, null)
>   voice of god("HI THERE", Emerald (/mob/living/silicon/ai), /list (/list), 1, 0, 1)
>   Voice of God (/obj/effect/proc_holder/spell/voice_of_god): cast(null, Emerald (/mob/living/silicon/ai))
>   Voice of God (/obj/effect/proc_holder/spell/voice_of_god): perform(null, 1, Emerald (/mob/living/silicon/ai))
>   Voice of God (/obj/effect/proc_holder/spell/voice_of_god): perform(null, 1, Emerald (/mob/living/silicon/ai))
>   Voice of God (/obj/effect/proc_holder/spell/voice_of_god): choose targets(Emerald (/mob/living/silicon/ai))
>   Voice of God (/obj/effect/proc_holder/spell/voice_of_god): Click()
>   Voice of God (/datum/action/spell_action): Trigger()
>   Voice of God (/obj/screen/movable/action_button): Click(null, "mapwindow.map", "icon-x=16;icon-y=6;left=1;scre...")

1) Be an AI.
2) Grant Voice of God spell.
3) Try using it, it runtimes.

@XDTM

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
Voice of god commands are now datums and allocated on a new .dm file, considering they amount to about 400+ lines.
Made the regexs a little more lenient with whitespaces overall. Also working on to fix #33650, but I still have to run a local test for which I don't have the time right now.
EDIT: Now it's ready.

## Why It's Good For The Game
Refactoring, not for the sake of it, but because that amount of else if is vexatious and probably not guidelines-friendly. This will close #33650.

## Changelog
:cl:
refactor: refactored the voice of god. Specifically, I made the commands into singleton datums. It's also more robust against edge cases and bad text matches.
fix: The voice of god spell should now work for AIs.
qol: the voice of god commands themselves should tolerate the lack of whitespaces now.
/:cl:

### CODE REVIEW

**Comment by MrMelbert** [comment]
`tweak` isn't a valid changelog prefix anymore, use `qol`

File: code/datums/voice_of_god_command.dm [hunk 1]
```diff
@@ -0,0 +1,427 @@
+#define COOLDOWN_STUN 1200
+#define COOLDOWN_DAMAGE 600
+#define COOLDOWN_MEME 300
+#define COOLDOWN_NONE 100
+
+GLOBAL_LIST_INIT(voice_of_god_commands, init_voice_of_god_commands())
+
+/proc/init_voice_of_god_commands()
+	. = list()
+	for(var/datum/voice_of_god_command/prototype in subtypesof(/datum/voice_of_god_command))
```
Comment (Mothblocks) [inline code/datums/voice_of_god_command.dm:10]:
```suggestion
	for(var/datum/voice_of_god_command/prototype as anything in subtypesof(/datum/voice_of_god_command))
```

File: code/datums/voice_of_god_command.dm [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+/datum/voice_of_god_command/sleeping/execute(list/listeners = list(), mob/living/user, power_multiplier = 1, message)
+	for(var/mob/living/carbon/target as anything in listeners)
+		target.Sleeping(40 * power_multiplier)
+
+/datum/voice_of_god_command/vomit
+	trigger = "vomit|throw\\s*up|sick"
+	cooldown = COOLDOWN_STUN
+
+/datum/voice_of_god_command/vomit/execute(list/listeners = list(), mob/living/user, power_multiplier = 1, message)
+	for(var/mob/living/carbon/target as anything in listeners)
```
Comment (Mothblocks) [inline code/datums/voice_of_god_command.dm:152]: > Ensure 'as anything' is not included for loops that don't target all living mob types.

🙂

File: code/datums/voice_of_god_command.dm [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+		trigger = regex(trigger)
+
+/*
+ * What happens when the command is triggered.
+ * Args:
+ * * listeners: the list of living mobs who are affected by the command.
+ * * user: the one who casted Voice of God
+ * * power_multiplier: multiplies the power of the command, most times.
+ */
+/datum/voice_of_god_command/proc/execute(list/listeners = list(), mob/living/user, power_multiplier = 1, message)
```
Comment (Mothblocks) [inline code/datums/voice_of_god_command.dm:119]: You shouldn't be doing `= list()` everywhere, you should just have your code work with null, or know that it'll never be null.

`for (var/name in null)` passes fine.

Comment (Mothblocks) [reply code/datums/voice_of_god_command.dm:119]: Same with everywhere else.

File: code/datums/voice_of_god_command.dm [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+/datum/voice_of_god_command/stun
+	trigger = "stop|wait|stand\\s*still|hold\\s*on|halt"
+	cooldown = COOLDOWN_STUN
+
+/datum/voice_of_god_command/stun/execute(list/listeners = list(), mob/living/user, power_multiplier = 1, message)
+	 // Ensure 'as anything' is not included for loops that don't target all living mob types.
+	for(var/mob/living/target as anything in listeners)
+		target.Stun(60 * power_multiplier)
+
+/datum/voice_of_god_command/paralyze
```
Comment (Mothblocks) [inline code/datums/voice_of_god_command.dm:131]: I think you should doc-comment every `/datum/voice_of_god_command`.

File: code/datums/voice_of_god_command.dm [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+	for(var/mob/living/target as anything in listeners)
+		target.heal_overall_damage(10 * power_multiplier, 10 * power_multiplier)
+
+/datum/voice_of_god_command/brute
+	trigger = "die|suffer|hurt|pain|death"
+	cooldown = COOLDOWN_DAMAGE
+
+/datum/voice_of_god_command/brute/execute(list/listeners = list(), mob/living/user, power_multiplier = 1, message)
+	for(var/mob/living/target as anything in listeners)
+		target.apply_damage(15 * power_multiplier, def_zone = BODY_ZONE_CHEST, wound_bonus=CANT_WOUND)
```
Comment (Mothblocks) [inline code/datums/voice_of_god_command.dm:194]: Be consistent with spacing.

File: code/datums/voice_of_god_command.dm [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+	for(var/mob/living/target as anything in listeners)
+		target.throw_at(get_step_towards(user, target), 3 * power_multiplier, 1 * power_multiplier)
+
+/datum/voice_of_god_command/who_are_you
+	trigger = "who\\s*are\\s*you|say\\s*your\\s*name|state\\s*your\\s*name|identify"
+
+/datum/voice_of_god_command/who_are_you/execute(list/listeners = list(), mob/living/user, power_multiplier = 1, message)
+	var/iteration = 1
+	for(var/mob/living/target as anything in listeners)
+		addtimer(CALLBACK(target, /atom/movable/proc/say, target.real_name), 5 * iteration)
```
Comment (Mothblocks) [inline code/datums/voice_of_god_command.dm:253]:
```suggestion
		addtimer(CALLBACK(target, /atom/movable/proc/say, target.real_name), 0.5 SECONDS * iteration)
```

Comment (Mothblocks) [reply code/datums/voice_of_god_command.dm:253]: Here and everywhere else.

File: code/datums/voice_of_god_command.dm [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+	for(var/mob/living/target as anything in listeners)
+		addtimer(CALLBACK(target, /atom/movable/proc/say, "Who's there?"), 5 * iteration)
+		iteration++
+
+/datum/voice_of_god_command/state_laws
+	trigger = "state\\s*(your)?\\s*laws"
+
+/datum/voice_of_god_command/state_laws/execute(list/listeners = list(), mob/living/user, power_multiplier = 1, message)
+	for(var/mob/living/silicon/target in listeners)
+		target.statelaws(force = 1)
```
Comment (Mothblocks) [inline code/datums/voice_of_god_command.dm:279]:
```suggestion
		target.statelaws(force = TRUE)
```

File: code/datums/voice_of_god_command.dm [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+	if(findtext(message, up_words))
+		direction = NORTH
+	else if(findtext(message, down_words))
+		direction = SOUTH
+	else if(findtext(message, left_words))
+		direction = WEST
+	else if(findtext(message, right_words))
+		direction = EAST
+	for(var/mob/living/target as anything in listeners)
+		addtimer(CALLBACK(GLOBAL_PROC, .proc/_step, target, direction? direction : pick(GLOB.cardinals)), 10 * (iteration - 1))
```
Comment (Mothblocks) [inline code/datums/voice_of_god_command.dm:300]:
```suggestion
		addtimer(CALLBACK(GLOBAL_PROC, .proc/_step, target, direction? direction : pick(GLOB.cardinals)), 1 SECONDS * (iteration - 1))
```

File: code/datums/voice_of_god_command.dm [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+			chair.buckle_mob(target)
+			break
+
+/datum/voice_of_god_command/stand
+	trigger = "stand"
+	is_regex = FALSE
+
+/datum/voice_of_god_command/stand/execute(list/listeners = list(), mob/living/user, power_multiplier = 1, message)
+	for(var/mob/living/target as anything in listeners)
+		if(target.buckled && istype(target.buckled, /obj/structure/chair))
```
Comment (Mothblocks) [inline code/datums/voice_of_god_command.dm:360]: Why bother checking that its a chair?

File: code/datums/voice_of_god_command.dm [hunk 10]
```diff
@@ -0,0 +1,427 @@
...
+		else
+			addtimer(CALLBACK(target, /mob/living/.proc/emote, "jump"), 5 * iteration)
+		iteration++
+
+/datum/voice_of_god_command/honk
+	trigger = "ho+nk"
+
+/datum/voice_of_god_command/honk/execute(list/listeners = list(), mob/living/user, power_multiplier = 1, message)
+	addtimer(CALLBACK(GLOBAL_PROC, .proc/playsound, get_turf(user), 'sound/items/bikehorn.ogg', 300, 1), 25)
+	if(user.mind && user.mind.assigned_role == "Clown")
```
Comment (Mothblocks) [inline code/datums/voice_of_god_command.dm:381]:
```suggestion
	if(user.mind?.assigned_role == "Clown")
```

File: code/datums/voice_of_god_command.dm [hunk 11]
```diff
@@ -0,0 +1,427 @@
+#define COOLDOWN_STUN 1200
+#define COOLDOWN_DAMAGE 600
+#define COOLDOWN_MEME 300
+#define COOLDOWN_NONE 100
```
Comment (Mothblocks) [inline code/datums/voice_of_god_command.dm:4]: Seconds.

File: code/datums/voice_of_god_command.dm [hunk 12]
```diff
@@ -0,0 +1,427 @@
...
+//////////////////////////////////////
+
+/*
+ * The main proc for the voice of god power. it makes the user shout a message in an ominous way,
+ * The first matching command (from a list of static datums) the listeners must obey,
+ * and the return value of this proc the cooldown variable of the command dictates. (only relevant for things with cooldowns i guess)
+ */
+/proc/voice_of_god(message, mob/living/user, list/span_list, base_multiplier = 1, include_speaker = FALSE, message_admins = TRUE)
+	var/log_message = uppertext(message)
+	if(!span_list || !span_list.len)
```
Comment (Mothblocks) [inline code/datums/voice_of_god_command.dm:25]:
```suggestion
	if(!LAZYLENGTH(span_list))
```

File: code/datums/voice_of_god_command.dm [hunk 13]
```diff
@@ -0,0 +1,427 @@
...
+			span_list = list("narsiesmall")
+		else
+			span_list = list()
+
+	if(!user.say(message, spans = span_list, sanitize = FALSE))
+		return
+
+	message = lowertext(message)
+	var/list/mob/living/listeners = list()
+	var/list/candidates = get_hearers_in_view(8, user) - (!include_speaker ? user : null)
```
Comment (Mothblocks) [inline code/datums/voice_of_god_command.dm:36]:
```suggestion
	var/list/candidates = get_hearers_in_view(8, user) - (include_speaker ? null : user)
```

Double negatives are confusing.

File: code/datums/voice_of_god_command.dm [hunk 14]
```diff
@@ -0,0 +1,427 @@
...
+	for(var/mob/living/L in candidates)
+		if(L.can_hear() && !L.anti_magic_check(FALSE, TRUE) && L.stat != DEAD)
```
Comment (Mothblocks) [inline code/datums/voice_of_god_command.dm:38]: Use named variables for the FALSE and TRUE if possible.

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Needs more seconds and such.

File: code/datums/voice_of_god_command.dm [hunk 15]
```diff
@@ -0,0 +1,427 @@
...
+	trigger = "dance"
+	is_regex = FALSE
+
+/datum/voice_of_god_command/emote/salute
+	trigger = "salute"
+	emote_name = "salute"
+	is_regex = FALSE
+
+/datum/voice_of_god_command/emote/play_dead
+	trigger = "play\\s*dead"
```
Comment (lordpidey) [inline code/datums/voice_of_god_command.dm:417]: While we're here, could you add "feign death"?

Comment (Mothblocks) [reply code/datums/voice_of_god_command.dm:417]: :-1: Make a different PR, refactors should not alter behavior.

Comment (lordpidey) [reply code/datums/voice_of_god_command.dm:417]: It already alters whitespace related behavior.

File: code/datums/voice_of_god_command.dm (hunk 2 above)
Comment (Ghommie) [reply code/datums/voice_of_god_command.dm:152]: Ah, good catch. 😛

File: code/datums/voice_of_god_command.dm [hunk 16]
```diff
@@ -294,16 +290,18 @@ GLOBAL_LIST_INIT(voice_of_god_commands, init_voice_of_god_commands())
...
 
 /// This command forces silicon listeners to state all their laws.
 /datum/voice_of_god_command/state_laws
 	trigger = "state\\s*(your)?\\s*laws"
 
 /datum/voice_of_god_command/state_laws/execute(list/listeners, mob/living/user, power_multiplier = 1, message)
+	var/iteration = 1
 	for(var/mob/living/silicon/target in listeners)
-		target.statelaws(force = 1)
+		addtimer(CALLBACK(target, /mob/living/silicon/proc/statelaws, TRUE), 0.5 SECONDS * iteration)
```
Comment (lordpidey) [inline code/datums/voice_of_god_command.dm:313]: Given that stating laws generally spans several messages, we'll probably want the delay per iteration a bit greater than half a second.

File: code/modules/mob/mob_helpers.dm [hunk 17]
```diff
@@ -248,6 +248,12 @@
 	firstname.Find(real_name)
 	return firstname.match
 
+/// Find the last name of a mob from the real name with regex
+/mob/proc/last_name()
+	var/static/regex/lasttname = new("\[^\\s-\]+$") //First word before whitespace or "-"
+	lasttname.Find(real_name)
+	return lasttname.match
```
Comment (lordpidey) [inline code/modules/mob/mob_helpers.dm:255]: Are you sure this didn't already exist?

I remember that heirlooms already were able to pull out the last name of "Lou Briccant"

File: code/datums/voice_of_god_command.dm [hunk 18]
```diff
@@ -32,11 +33,36 @@ GLOBAL_LIST_INIT(voice_of_god_commands, init_voice_of_god_commands())
...
+	var/list/specific_listeners = list()
+	var/found_string = null
+
 	var/list/mob/living/listeners = list()
-	var/list/candidates = get_hearers_in_view(8, user) - (!include_speaker ? user : null)
-	for(var/mob/living/L in candidates)
-		if(L.can_hear() && !L.anti_magic_check(FALSE, TRUE) && L.stat != DEAD)
-			listeners += L
+	var/list/candidates = get_hearers_in_view(8, user) - (include_speaker ? null : user)
+	for(var/mob/living/living in candidates)
```
Comment (Mothblocks) [inline code/datums/voice_of_god_command.dm:53]: `living` is still a bad name here. Just make it `candidate`. 😄

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/modules/mob/mob_helpers.dm (hunk 17 above)
Comment (Ghommie) [reply code/modules/mob/mob_helpers.dm:255]: the heirloom component just does something along the lines of `var/list/names = splittext(name, " ")` followed by **var/surname = names[names.len]**. I feel like this one is more consistent with `first_name` and also safer to use on hyphenated names.

**Pull request merged**